RFC-0006-nan-definition.md

```diff
@@ -0,0 +1,179 @@
+
+### IEEE-754 with total order
+
+In this option we implement the full IEEE spec including the total order definition to use for sorting and order by.  NaN = NaN would be false.  NaN less than or greater than anything is false and NaN <> NaN is true. Group bys and distincts would return separate groups for each NaN value.  Order by on a double column containing NaN would give you the IEEE defined total order.  An open question would be the behavior of functions like min or max (are they functions on NaN input that should return NaN or are they sorting functions that should return based on the total order? what about when it's a topn function like min(x, n) or max(x, n)?).  
+
+Advantages here are full compliance with the IEEE spec.
+
+Disadvantages are the inconsistency between comparison and sorting operators can lead to confusion and bugs. It keeps the funny behavior with maps where you can put nans in but can't get them out (and not sure if it would cause implementation problems if we did group bys this way). It does not seem like other dbs do this either.
+
+### Disallow NaNs
+
+In this option we would throw an exception if a NaN value would be encountered. We might then want to do the same for Infinity and - Infinity as the logic would be the same. 
+
+This option might be in greater compliance with the SQL spec, which does not specify the special constants NaN, infinity, and -infinity). This approach also ensures that the user never misuses NaNs in their data due to not understanding how Presto handles them. 
+
+A big disadvantage here is that Presto reads tables that were written by other systems that do support NaN. We need to be able to read those files and not return an error. We could have some kind of function that does some special handling if you ask it to, but the whole thing would be a bit convoluted. 
+s
+
+## Adoption Plan
+
+- Adopting this definition will require changing behavior for many Presto functions, For some the NaN behavior is a bit surprising already and should not need a special flag (e.g. you can put as many NaN keys in a map as you want, but can never get them out again).  Others are basic operations like `<, >, =, <>` or joins, and sudden changes in these behaviors could cause problems for some users.
+- We will have a temporary configuration/session property ``use_old_nan_behavior`` for changes to these basic functions to help with the transition. This property will be removed after a couple releases and we will switch fully to the new definition
+- We will add thorough documentation for the new behavior..
+
+## Test Plan
+
+We will add a test class TestNanQueries with tests to ensure correct behavior for all functions/operations/etc. that handle NaNs. The tests will use queries that execute on workers, rather than constants that get interpreted on the coordinator. That way the test can also be used to ensure consistency between native and java workers.
```

Comment:
```suggestion
We will add a test class TestNanQueries with tests to ensure correct behavior for all functions and operations that handle NaNs. The tests will use queries that execute on workers, rather than constants that get interpreted on the coordinator, so that the test can also be used to ensure consistency between native and java workers.
```